extern/boostd-data/couchbase/db.go

```diff
@@ -310,136 +320,229 @@ func (db *DB) AddDealForPiece(ctx context.Context, pieceCid cid.Cid, dealInfo mo
+			if !ok {
+				return nil, fmt.Errorf("parsing piece %s offset value '%s': missing Offset map key", pieceCid, val)
+			}
+			offsetStr, ok := offsetIfce.(string)
+			if !ok {
+				return nil, fmt.Errorf("unexpected type for piece cid %s offset value: %T", pieceCid, offsetIfce)
+			}
+			offset, err := strconv.ParseUint(offsetStr, 10, 64)
+			if err != nil {
+				return nil, fmt.Errorf("parsing piece %s offset value '%s' as uint64: %w", pieceCid, val, err)
+			}
+
+			sizeIfce, ok := val["s"]
+			if !ok {
+				return nil, fmt.Errorf("parsing piece %s offset value '%s': missing Size map key", pieceCid, val)
+			}
+			sizeStr, ok := sizeIfce.(string)
+			if !ok {
+				return nil, fmt.Errorf("unexpected type for piece cid %s size value: %T", pieceCid, offsetIfce)
+			}
+			size, err := strconv.ParseUint(sizeStr, 10, 64)
+			if err != nil {
+				return nil, fmt.Errorf("parsing piece %s size value '%s' as uint64: %w", pieceCid, val, err)
+			}
+
+			recs = append(recs, model.Record{Cid: cid.NewCidV1(cid.Raw, mh), OffsetSize: model.OffsetSize{
+				Offset: offset,
+				Size:   size,
+			}})
+		}
 	}
 
 	return recs, nil
 }
 
 type offsetSize struct {
 	// Need to use strings instead of uint64 because uint64 may not fit into
-	// Javascript number
+	// a Javascript number
 	Offset string `json:"o"`
 	Size   string `json:"s"`
 }
 
+// Couchbase has an upper limit on the size of a value: 20mb
+// A JSON-encoded map with 128k keys results in a value of about 8mb in size
+// so this is well under the 20mb limit.
+var maxRecsPerShard = 128 * 1024
+
+// Limit the number of shards to 2048. This means there is an upper limit of
+// ~270 million blocks per piece, which should be more than enough:
+// 64 Gib piece / (2048 * 128 * 1024) = 238 bytes per block
+const maxShardsPerPiece = 2048
+
+var maxRecsPerPiece = maxShardsPerPiece * maxRecsPerShard
+
+func getShardPrefixBitCount(recordCount int) (int, int) {
+	// The number of shards required to store all the keys
+	requiredShards := (recordCount / maxRecsPerShard) + 1
+	// The number of shards that will be created must be a power of 2,
+	// so get the first power of two that's >= requiredShards
+	shardPrefixBits := 0
+	totalShards := 1
+	for totalShards < requiredShards {
+		shardPrefixBits += 1
+		totalShards *= 2
+	}
+
+	return shardPrefixBits, totalShards
+}
+
+func getShardPrefix(shardIndex int) string {
+	shardPrefix := []byte{0, 0}
+	shardPrefix[1] = byte(shardIndex)
+	shardPrefix[0] = byte(shardIndex >> 8)
+	return string(shardPrefix)
+}
+
 // AddIndexRecords
 func (db *DB) AddIndexRecords(ctx context.Context, pieceCid cid.Cid, recs []model.Record) error {
 	ctx, span := tracing.Tracer.Start(ctx, "db.add_index_records")
 	defer span.End()
 
-	mhToOffsetSize := make(map[string]offsetSize)
-	for _, rec := range recs {
-		mhToOffsetSize[rec.Cid.Hash().String()] = offsetSize{
-			Offset: fmt.Sprintf("%d", rec.Offset),
-			Size:   fmt.Sprintf("%d", rec.Size),
-		}
+	if len(recs) > maxRecsPerPiece {
```

Comment:
We should be logging / emitting the `recs` value, so that we have an idea of how large various indices are.